### PR TITLE
Warm instances

### PR DESCRIPTION
**Pull request by ovstinga**
- [x] use warm instances
- [x] unit tests
- [x] mem leaks fixes
- [x] profiling
- [x] instance `reset` on **wasmer** to sync
- [x] integration tests
- [x] refactor

### CODE REVIEW

**Comment by ovstinga** [comment]
TODO: 

- [ ] add fuzzy test from another branch (already created)
- [ ] profile again with different values to check for any memory leaks (just to be extra sure...)
- [ ] small cleanup (refactor)

File: arwen/contexts/blockchain_test.go [hunk 1]
```diff
@@ -2,13 +2,14 @@ package contexts
 
 import (
 	"errors"
+	"github.com/ElrondNetwork/wasm-vm/crypto/hashing"
```
Comment (sasurobert) [inline arwen/contexts/blockchain_test.go:5]: goimports

File: arwen/contexts/runtime.go [hunk 2]
```diff
@@ -135,11 +133,17 @@ func (context *runtimeContext) StartWasmerInstance(contract []byte, gasLimit uin
...
+	warmInstanceUsed := context.useWarmInstanceIfExists(gasLimit, newCode)
+	if warmInstanceUsed {
+		return nil
+	}
 	compiledCodeUsed := context.makeInstanceFromCompiledCode(gasLimit, newCode)
 	if compiledCodeUsed {
 		return nil
 	}
 
+	//fmt.Println("byteCodeUsed")
```
Comment (sasurobert) [inline arwen/contexts/runtime.go:146]: delete

File: arwen/contexts/runtime.go [hunk 3]
```diff
@@ -859,6 +928,16 @@ func (context *runtimeContext) CleanInstance() {
 	return
 }
 
+func (context *runtimeContext) ShallowClean() {
```
Comment (sasurobert) [inline arwen/contexts/runtime.go:931]: add comment to unexported function.

File: mock/context/instanceMock.go [hunk 4]
```diff
@@ -150,6 +150,13 @@ func GetMockInstance(host arwen.VMHost) *InstanceMock {
 	return instance
 }
 
+func (instance *InstanceMock) ShallowCopy() wasmer.InstanceHandler {
```
Comment (sasurobert) [inline mock/context/instanceMock.go:153]: comment on public functions

File: mock/context/instanceMock.go [hunk 5]
```diff
@@ -150,6 +150,13 @@ func GetMockInstance(host arwen.VMHost) *InstanceMock {
...
+	return instance
+}
+
+func (instance *InstanceMock) ShallowClean() {
```
Comment (sasurobert) [inline mock/context/instanceMock.go:157]: comment on public functions

File: mock/world/worldAccountMap.go [hunk 6]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/wasm-vm/crypto/hashing"
```
Comment (sasurobert) [inline mock/world/worldAccountMap.go:7]: goimports

File: wasmer/instance.go [hunk 7]
```diff
@@ -265,6 +267,38 @@ func (instance *Instance) Clean() {
...
 		}
 	}
+
+	instance.Data = nil
+	instance.DataPointer = nil
+	instance.Exports = nil
+	instance.Signatures = nil
+}
+
+func (instance *Instance) ShallowClean() {
```
Comment (sasurobert) [inline wasmer/instance.go:277]: add comment to public function

File: wasmer/instance.go [hunk 8]
```diff
@@ -265,6 +267,38 @@ func (instance *Instance) Clean() {
...
+	instance.Memory = nil
+	instance.Data = nil
+	instance.DataPointer = nil
+	instance.Exports = nil
+	instance.Signatures = nil
+}
+
```
Comment (sasurobert) [inline wasmer/instance.go:284]: add comment to public function

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**